src/cryptography/hazmat/backends/openssl/aead.py

```diff
@@ -111,12 +134,29 @@ def _decrypt(backend, cipher_cls, key, nonce, data, associated_data,
     ctx = _aead_setup(
         backend, cipher_name, key, nonce, tag, tag_length, _DECRYPT
     )
+    # CCM requires us to pass the length of the data before processing anything
+    from cryptography.hazmat.primitives.ciphers.aead import AESCCM
+    if cipher_cls is AESCCM:
+        _set_length(backend, ctx, len(data))
+
     _process_aad(backend, ctx, associated_data)
-    processed_data = _process_data(backend, ctx, data)
-    outlen = backend._ffi.new("int *")
-    res = backend._lib.EVP_CipherFinal_ex(ctx, backend._ffi.NULL, outlen)
-    if res == 0:
-        backend._consume_errors()
-        raise InvalidTag
+    # CCM has a different error path if the tag doesn't match. Errors are
+    # raised in Update and Final is irrelevant.
+    if cipher_name[-3:] == b"ccm":
+        outlen = backend._ffi.new("int *")
+        buf = backend._ffi.new("unsigned char[]", len(data))
+        res = backend._lib.EVP_CipherUpdate(ctx, buf, outlen, data, len(data))
```

Comment:
Because we need to check the error code and `_process_data` asserts on the error code and returns the buffer. Plus, calling EVP_CipherFinal returns zero with CCM (what a horribly overloaded API EVP is)